src/algorithms/reco/MatchClusters.cc

```diff
@@ -14,197 +15,170 @@
+            debug("    --> adding cluster to reconstructed particle");
+            outpart.addToClusters(clus);
+            clusterMap.erase(mcID);
         }
 
-        return {std::move(outparts), std::move(outpartsassoc)};
+        // create truth associations
+        auto assoc = outpartsassoc->create();
+        assoc.setRecID(outpart.getObjectID().index);
+        assoc.setSimID(mcID);
+        assoc.setWeight(1.0);
+        assoc.setRec(outpart);
+        assoc.setSim((*mcparticles)[mcID]);
     }
 
+    // 2. Now loop over all remaining clusters and add neutrals. Also add in Hcal energy
+    // if a matching cluster is available
+    debug("Step 2/2: Creating neutrals for remaining clusters...");
+    for (const auto &[mcID, clus]: clusterMap) {
+        debug(" --> Processing unmatched cluster with energy: {}", clus.getEnergy());
+
+
+        // get mass/PDG from mcparticles, 0 (unidentified) in case the matched particle is charged.
+        const auto mc = (*mcparticles)[mcID];
+        const double mass = (!mc.getCharge()) ? mc.getMass() : 0;
+        const int32_t pdg = (!mc.getCharge()) ? mc.getPDG() : 0;
+        if (level() <= algorithms::LogLevel::kDebug) {
+            if (mc.getCharge()) {
+                debug("   --> associated mcparticle is not a neutral (PDG: {}), "
+                                "setting the reconstructed particle ID to 0 (unidentified)", mc.getPDG());
+            }
+            debug("   --> found matching associated mcparticle with PDG: {}, energy: {}", pdg,
+                            mc.getEnergy());
+        }
 
+        // Reconstruct our neutrals and add them to the list
+        const auto outpart = reconstruct_neutral(&clus, mass, pdg);
+        debug(" --> Reconstructed neutral particle with PDG: {}, energy: {}", outpart.getPDG(),
+                        outpart.getEnergy());
 
-    // get a map of mcID --> cluster
-    // input: cluster_collections --> list of handles to all cluster collections
-    std::map<int, edm4eic::Cluster> MatchClusters::indexedClusters(
-            const std::vector<const edm4eic::ClusterCollection*> &cluster_collections,
-            const std::vector<const edm4eic::MCRecoClusterParticleAssociationCollection*> &associations_collections) {
-        std::map<int, edm4eic::Cluster> matched = {};
+        outparts->push_back(outpart);
 
-        // loop over cluster collections
-        for (const auto &clusters: cluster_collections) {
+        // Create truth associations
+        auto assoc = outpartsassoc->create();
+        assoc.setRecID(outpart.getObjectID().index);
+        assoc.setSimID(mcID);
+        assoc.setWeight(1.0);
+        assoc.setRec(outpart);
+        assoc.setSim((*mcparticles)[mcID]);
+    }
+}
 
-            // loop over clusters
-            for (const auto cluster: *clusters) {
+// get a map of mcID --> cluster
+// input: clusters --> all clusters
+std::map<int, edm4eic::Cluster> MatchClusters::indexedClusters(
```

Comment:
```suggestion
std::map<edm4eic::MCParticle, edm4eic::Cluster> MatchClusters::indexedClusters(
```

Not sure on this one but it might be nicer to pass around the reference to the MCParticle directly rather than the ID within the collection. Would need to be altered in a few placed.